pinot-core/src/main/java/org/apache/pinot/core/transport/DirectOOMServerHandler.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.pinot.common.metrics.ServerMeter;
+import org.apache.pinot.common.metrics.ServerMetrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handling netty direct memory OOM. In this case there is a great chance that servers are receiving requests from
+ * broker concurrently that needs lot of scans on server pusing servers to OOM. We want to close all channels to broker
+ * to proactively release the direct memory.
+ */
+public class DirectOOMServerHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DirectOOMServerHandler.class);
+  private static final AtomicBoolean OOM_SHUTTING_DOWN = new AtomicBoolean(false);
+
+  private final ConcurrentHashMap<SocketChannel, Boolean> _allChannels;
+  private volatile boolean _silentShutDown = false;
+  private ServerSocketChannel _serverSocketChannel;
+
+  public DirectOOMServerHandler(ConcurrentHashMap<SocketChannel, Boolean> allChannels,
+      ServerSocketChannel serverSocketChannel) {
+    _allChannels = allChannels;
+    _serverSocketChannel = serverSocketChannel;
+  }
+
+  public void setSilentShutDown() {
+    _silentShutDown = true;
+  }
+
+  /**
+   * Closes and removes all active channels from the map to release direct memory.
+   */
+  private void closeAllChannels() {
+    LOGGER.warn("OOM detected: Closing all channels to release direct memory");
+    for (SocketChannel channel : _allChannels.keySet()) {
+      try {
+        channel.close();
+      } catch (Exception e) {
+        LOGGER.error("Error while closing channel: {}", channel, e);
+      } finally {
+        _allChannels.remove(channel);
+      }
+    }
+  }
+
+  void setSilentShutdown() {
+    if (_serverSocketChannel != null) {
+      DirectOOMServerHandler directOOMHandler = _serverSocketChannel.pipeline().get(DirectOOMServerHandler.class);
+      if (directOOMHandler != null) {
+        directOOMHandler.setSilentShutDown();
+      }
+    }
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    // if we are shutting down channels due to direct memory OOM, we short circuit the channel inactive
+    if (_silentShutDown) {
+      return;
+    }
+    ctx.fireChannelInactive();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+    if (cause instanceof OutOfMemoryError && cause.getMessage().contains("direct buffer")) {
+      ServerMetrics.get().addMeteredGlobalValue(ServerMeter.DIRECT_MEMORY_OOM, 1L);
+      // Only one thread should handle OOM
+      if (OOM_SHUTTING_DOWN.compareAndSet(false, true)) {
+        try {
+          LOGGER.error("Direct memory OOM detected. Closing all channels to free up memory.", cause);
+          closeAllChannels();
+          setSilentShutdown();
+        } catch (Exception e) {
+          LOGGER.error("Error while handling direct memory OOM", e);
+        } finally {
+          OOM_SHUTTING_DOWN.set(false);
+        }
```

Comment:
What happens with new request once this code is fired? It looks to me that this handler is kept in the pipeline, so we will add new channels into _allChannels, right? But then _silentShutdown is kept as true forever. That sounds incorrect, although I may not be correctly understanding the flow